File: HugeInteger.cc
#include "HugeInteger.h"
#include <math.h>
#include <iostream>
#include <stdio.h>
#include <vector>
using std::vector;
#include <sstream>
using std::stringstream;
#include <string>
using std::string;

//Initializes HugeInteger to zero
HugeInteger::HugeInteger() {digits.push_back(0);};

//Makes a copy of HugeInteger
HugeInteger::HugeInteger(const HugeInteger & other) {
  for(int i = 0; i <= other.digits.size(); i++){
      digits.push_back(other.digits[i]);
  }
}

//Initializes HugeInteger to value contained int the string
HugeInteger::HugeInteger(const string & s) {
  for(int i = 0; i<s.length(); i++) {
    //Checks if the character is a number and that it would not be a leading zero
    if(s[i] >= '0' && s[i] <= '9' && !(digits.size() == 0 && s[i] == '0')) {
      int x = s[i] - '0';
      digits.insert(digits.begin(), x);

    }
  }
  //If there is nothing in the vector, it pushes a 0 onto it
  if(digits.empty()) {
    digits.push_back(0);
  }
}

//Overloads the == operator so that it works with the HugeIntegers
bool HugeInteger::operator==(const HugeInteger & rightSide) {
  //Checks if the HugeIntegers are different sizes
  if(this->digits.size() > rightSide.digits.size() || this->digits.size() > rightSide.digits.size()) {
    return false;
  }
  //Checks if each number of the HugeIntegers are the same
  else {
    for(int i = rightSide.digits.size() - 1; i >= 0; i--) {
      if(rightSide.digits[i] > this->digits[i] || this->digits[i] > rightSide.digits[i]) {
        return false;
      }
    }
  }
  return true;
}

//Overloads the + operator so it can add two HugeIntegers
const HugeInteger HugeInteger::operator+(const HugeInteger & rightSide) {
  HugeInteger HTemp = HugeInteger();
  int remain = 0;
  HTemp.digits[0] = this->digits[0] + rightSide.digits[0];
  if(this->digits[0] + rightSide.digits[0] > 9) {
    HTemp.digits[0] = HTemp.digits[0] - 10;
    remain = 1;
  }
  //Checks if this is larger than rightSide
  if(this->digits.size() > rightSide.digits.size()) {
    for(int i = 1; i < rightSide.digits.size(); i++) {
      HTemp.digits.push_back(this->digits[i] + rightSide.digits[i] + remain);
      //Checks if the addition has gone over 9 into double digits and creates a remainder placeholder
      if(this->digits[i] + rightSide.digits[i] + remain > 9) {
        HTemp.digits[i] = HTemp.digits[i] - 10;
        remain = 1;
      }
      else {
        remain = 0;
      }
    }
    //Finishes the number with the remaining numbers of this
    if(HTemp.digits.size() < this->digits.size()) {
      for(int i = rightSide.digits.size(); i < this->digits.size(); i++) {
        HTemp.digits.push_back(this->digits[i] + remain);
        remain = 0;
      }
    }
  }
    //Same as above but for the cases that rightSide is greater than this or they are equal
    else {
      for(int i = 1; i < this->digits.size(); i++) {
        HTemp.digits.push_back(this->digits[i] + rightSide.digits[i] + remain);
        if(this->digits[i] + rightSide.digits[i] + remain > 9) {
          HTemp.digits[i] = HTemp.digits[i] - 10;
          remain = 1;
        }
        else {
          remain = 0;
        }
      }
      if(HTemp.digits.size() < rightSide.digits.size()) {
        for(int i = this->digits.size(); i < rightSide.digits.size(); i++) {
          HTemp.digits.push_back(rightSide.digits[i] + remain);
          remain = 0;
      }
    }
  }
  if(remain == 1) {
    HTemp.digits.push_back(1);
  }
  return HTemp;
}


//Overloads the > operator so that it can compare two HugeIntegers
bool HugeInteger::operator>(const HugeInteger & rightSide) {
  //Checks if one HugeInteger is larger than the other
  if(this->digits.size() > rightSide.digits.size()) {
    return true;
  }
  else if(this->digits.size() < rightSide.digits.size()) {
    return false;
  }
  //Cycles throught the digits of the HugeIntegers and checks which is larger
  else {
    for(int i = rightSide.digits.size() - 1; i >= 0; i--) {
      if(rightSide.digits[i] > this->digits[i]) {
        return false;
      }
      else if(rightSide.digits[i] < this->digits[i]) {
        return true;
      }
    }
  }
  return false;
}
//Checks if the given HugeInteger is zero
bool HugeInteger::isZero() const {
  for(int i = 0; i < this->digits.size(); i++) {
    if(this->digits[i] != 0) {
      return false;
    }
  }
  return true;
}

//Returns a string representation of the HugeInteger
string HugeInteger::toString() const {
  stringstream stream;
  for(int i = digits.size() - 1; i >= 0; i--) {
    stream << digits[i];
  }

  string result = stream.str();
  return result;
}

File: HugeInteger.go
/*
* NAMES: Charlie LeWarne, Adrian Ronquillo
* CLASS: CSCI 343
* PROFESSOR: Dr. Murphy
* DATE: May 11, 2020
* SOURCES: All contained in the slides
* Main source used was golang.org and our own work from past semesters in C++ and Java
* SPECIAL INSTRUCTIONS: None
 */

package main

import (
	"fmt"
	"strconv"
)

/*
* HugeInteger struct which holds the data in the HugeInteger
 */
type HugeInteger struct {
	numberholder []int
}

/*
* Initializes the HugeInteger to h1, is the copy constructor
 */
func HugeIntegerC(h1 HugeInteger) HugeInteger {
	var output []int
	output = h1.numberholder

	h2 := HugeInteger{
		numberholder: output}

	return h2
}

/*
* Initializes the HugeInteger to 0, is the constructor for an empty HugeInteger
 */
func HugeIntegerD() HugeInteger {
	var temp []int
	temp = append(temp, 0)
	h := HugeInteger{
		numberholder: temp}
	return h
}

/*
* Initializes the HugeInteger to the input integer
 */
func HugeIntegerN(input int) HugeInteger {
	var temp []int
	s := strconv.Itoa(input)
	for i := len(s) - 1; i >= 0; i-- {
		j, err := strconv.Atoi(string(s[i])) //err checks to see if the number is truly a number
		if err == nil {                      //if it is a number, this gets appended to the numberholder
			temp = append(temp, j)
		}
	}
	h := HugeInteger{
		numberholder: temp}
	return h
}

/*
* Initializes the HugeInteger to the input string
* Do not assume there are not leading zeroes. Octal problem is solved when passing a string
 */
func HugeIntegerS(input string) HugeInteger {
	var temp []int
	var ZF int
	var output []int
	for i := len(input) - 1; i >= 0; i-- { //String is iterated through to add each element to a temporary slice
		j, err := strconv.Atoi(string(input[i]))
		if err == nil {
			if j == 0 {
				ZF++
			} else {
				ZF = 0
			}
			temp = append(temp, j)
		}
	}
	if ZF != 0 { //ZF is created to count the number of leading zeroes
		output = temp[:len(temp)-ZF]
	} else {
		output = temp
	}
	h := HugeInteger{ //HugeInteger is created with the temporary slice being used as the numberholder
		numberholder: output}
	return h
}

/*
* Function to add two HugeIntegers together and to return a third, new HugeInteger
 */
func add(h1, h2 HugeInteger) HugeInteger {
	var temp []int
	var num int
	var rem int
	rem = 0
	if len(h1.numberholder) > len(h2.numberholder) {
		for i := 0; i < len(h2.numberholder); i++ {
			num = h2.numberholder[i] + h1.numberholder[i] + rem
			if num >= 10 {
				num = num - 10
				rem = 1
			} else {
				rem = 0
			}
			temp = append(temp, num)
		}
		for i := len(h2.numberholder); i < len(h1.numberholder); i++ {
			if rem != 1 {
				num = h1.numberholder[i]
			} else {
				num = h1.numberholder[i] + 1
				if num >= 10 {
					num = num - 10
					rem = 1
				} else {
					rem = 0
				}
			}
			temp = append(temp, num)
		}
	} else {
		for i := 0; i < len(h1.numberholder); i++ {
			num = h2.numberholder[i] + h1.numberholder[i] + rem
			if num >= 10 {
				num = num - 10
				rem = 1
			} else {
				rem = 0
			}
			temp = append(temp, num)
		}
		for i := len(h1.numberholder); i < len(h2.numberholder); i++ {
			if rem != 1 {
				num = h2.numberholder[i]
			} else {
				num = h2.numberholder[i] + 1
				rem = 0
			}
			temp = append(temp, num)
		}
	}
	if rem == 1 {
		temp = append(temp, 1)
	}
	h := HugeInteger{
		numberholder: temp}
	return h
}

/*
* Prints out each element in the HugeInteger inputted as the reciever
 */
func toString(h HugeInteger) string {
	var returnString string = ""
	for i := len(h.numberholder) - 1; i >= 0; i-- {
		returnString += strconv.Itoa(h.numberholder[i])
	}
	return returnString
}

/*
* Checks this HugeInteger to see if it contains all zeros
 */
func isZero(h HugeInteger) bool {
	for i := len(h.numberholder) - 1; i >= 0; i-- {
		if h.numberholder[i] != 0 { //if a number is found that isn't 0, then this method returns false
			return false
		}
	}
	return true //if the the hugeinteger is all zeros
}

/*
* Compares two HugeIntegers to see if they are equal
 */
func equals(h1 HugeInteger, h2 HugeInteger) bool {
	if len(h1.numberholder) != len(h2.numberholder) {
		return false
	} //checks to see if the sizes of the hugeintegers are equal, if not then throws false
	for i := 0; i < len(h1.numberholder); i++ { //if the sizes of the hugeintegers are equal
		if h1.numberholder[i] != h2.numberholder[i] {
			return false
		}
	}
	return true //the hugeintegers are equal
}

/*
* Compares two HugeIntegers h1 and h2. If h1 is larger than h2 then true is returned, false otherwise
 */
func greaterThan(h1 HugeInteger, h2 HugeInteger) bool {
	if len(h1.numberholder) > len(h2.numberholder) {
		return true
	} else if len(h1.numberholder) < len(h2.numberholder) {
		return false
	}
	if len(h1.numberholder) == 1 {
		if h1.numberholder[0] > h2.numberholder[0] {
			return true
		} else if h1.numberholder[0] < h2.numberholder[0] {
			return false
		}
		return false
	}
	for i := len(h1.numberholder) - 1; 0 <= i; i-- { //if the lengths of both hugeintegers are equal
		if h1.numberholder[i] > h2.numberholder[i] {
			return true
		} else if h1.numberholder[i] < h2.numberholder[i] {
			return false
		}
	}
	return false //if the above checks do not return anything then the two hugeintegers are equal to eachother

}

/*
* Compares two HugeIntegers h1 and h2. If h2 is larger than h1 then true is returned, false otherwise
 */
func lessThan(h1 HugeInteger, h2 HugeInteger) bool {
	if len(h1.numberholder) < len(h2.numberholder) {
		return true
	} else if len(h1.numberholder) > len(h2.numberholder) {
		return false
	}
	if len(h1.numberholder) == 1 {
		if h1.numberholder[0] < h2.numberholder[0] {
			return true
		} else if h1.numberholder[0] > h2.numberholder[0] {
			return false
		}
		return false
	}
	for i := len(h1.numberholder) - 1; 0 <= i; i-- { //if the lengths of both hugeintegers are equal
		if h1.numberholder[i] < h2.numberholder[i] {
			return true
		} else if h1.numberholder[i] > h2.numberholder[i] {
			return false
		}
	}
	return false //if the above checks do not return anything then the two hugeintegers are equal to eachother

}

func main() {
	a := HugeIntegerD()
	b := HugeIntegerN(9999)
	c := HugeIntegerS("1")
	d := add(b, c)
	e := isZero(a)
	f := equals(b, c)
	g := greaterThan(b, c)
	i := lessThan(b, c)
	var aa string = toString(a)
	var bb string = toString(b)
	var cc string = toString(c)
	var dd string = toString(d)
	fmt.Println(aa)
	fmt.Println(bb)
	fmt.Println(cc)
	fmt.Println(dd)
	fmt.Println(e)
	fmt.Println(f)
	fmt.Println(g)
	fmt.Println(i)
}
